plugins/usbdmx/AsyncUsbReceiver.cpp

```diff
@@ -0,0 +1,180 @@
+  if (!m_usb_handle) {
+    OLA_WARN << "AsyncUsbReceiver hasn't been initialized";
+    return false;
+  }
+  ola::thread::MutexLocker locker(&m_mutex);
+  return PerformTransfer();
+}
+
+void AsyncUsbReceiver::CancelTransfer() {
+  if (!m_transfer) {
+    return;
+  }
+
+  bool canceled = false;
+  while (1) {
+    ola::thread::MutexLocker locker(&m_mutex);
+    if (m_transfer_state == IDLE || m_transfer_state == DISCONNECTED) {
+      break;
+    }
+    if (!canceled) {
+      m_suppress_continuation = true;
+      if (m_adaptor->CancelTransfer(m_transfer) == 0) {
+        canceled = true;
+      } else {
+        break;
+      }
+    }
+  }
+
+  m_suppress_continuation = false;
+}
+
+void AsyncUsbReceiver::FillControlTransfer(unsigned char *buffer,
+                                           unsigned int timeout) {
+  m_adaptor->FillControlTransfer(m_transfer, m_usb_handle, buffer,
+                                 &AsyncCallback, this, timeout);
+}
+
+void AsyncUsbReceiver::FillBulkTransfer(unsigned char endpoint,
+                                        unsigned char *buffer,
+                                        int length,
+                                        unsigned int timeout) {
+  m_adaptor->FillBulkTransfer(m_transfer, m_usb_handle, endpoint, buffer,
+                              length, &AsyncCallback, this, timeout);
+}
+
+void AsyncUsbReceiver::FillInterruptTransfer(unsigned char endpoint,
+                                             unsigned char *buffer,
+                                             int length,
+                                             unsigned int timeout) {
+  m_adaptor->FillInterruptTransfer(m_transfer, m_usb_handle, endpoint, buffer,
+                                   length, &AsyncCallback, this, timeout);
+}
+
+int AsyncUsbReceiver::SubmitTransfer() {
+  int ret = m_adaptor->SubmitTransfer(m_transfer);
+  if (ret) {
+    OLA_WARN << "libusb_submit_transfer returned "
+             << m_adaptor->ErrorCodeToString(ret);
+    if (ret == LIBUSB_ERROR_NO_DEVICE) {
+      m_transfer_state = DISCONNECTED;
+    }
+    return false;
+  }
+  m_transfer_state = IN_PROGRESS;
+  return ret;
+}
+
+void AsyncUsbReceiver::TransferComplete(struct libusb_transfer *transfer) {
+  if (transfer != m_transfer) {
+    OLA_WARN << "Mismatched libusb transfer: " << transfer << " != "
+             << m_transfer;
+    return;
+  }
+
+  if (transfer->status != LIBUSB_TRANSFER_COMPLETED &&
+      transfer->status != LIBUSB_TRANSFER_TIMED_OUT ) {
+    OLA_WARN << "Transfer returned " << transfer->status;
+  }
+
+  ola::thread::MutexLocker locker(&m_mutex);
+  m_transfer_state = (transfer->status == LIBUSB_TRANSFER_NO_DEVICE ?
+      DISCONNECTED : IDLE);
+
+  if (m_suppress_continuation) {
+    return;
+  }
+
+  if (transfer->status != LIBUSB_TRANSFER_TIMED_OUT) {
+    if (TransferCompleted(&m_rx_buffer)) {  // Input changed
+      if (m_receive_callback.get()) {
+        m_receive_callback->Run();
```

Comment:
This suffers from the same issue, the callback is getting run in a thread which isn't the main event thread.

However if you run the callback in the main thread, you'll then have the problem of calling GetDmxInBuffer() and accessing m_rx_buffer without a lock. 

I think what you want to do it:
 - receive the completion event in the libusb thread
 - call TransferComplete to unpack the response into a DMXBuffer
 - Schedule a callback on the Executor to run in the main thread

Now there are two ways to transfer the DMX data:
i) Have the main thread call GetDMX(DMXBuffer *dmx), the method takes a lock, copies the data from the receiver's buffer to the argument buffer and returns.
ii) Change the RX callback to take a DMXBuffer*, in which case the receiver would allocate a new DMX buffer for each frame and then pass off ownership of the DMXBuffer to whatever receives the callback. 

i) involves a mutex acquisition & a memcpy, ii) involves 2 calls to new (one for the DMXBuffer & one for the outer callback binding to the DMXBuffer) & two calls to delete.